xdsl/transforms/apply_pdl.py

```diff
@@ -0,0 +1,43 @@
+import os
+from dataclasses import dataclass
+from io import StringIO
+from typing import cast
+
+from xdsl.context import MLContext
+from xdsl.dialects import builtin, pdl
+from xdsl.interpreters.pdl import (
+    PDLRewritePattern,
+)
+from xdsl.parser import Parser
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriteWalker,
+    RewritePattern,
+)
+
+
+@dataclass(frozen=True)
+class ApplyPDLPass(ModulePass):
+    name = "apply-pdl"
+
+    pdl_file: str | None = None
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        payload_module = op
+        if self.pdl_file is not None:
+            assert os.path.exists(self.pdl_file)
+            with open(self.pdl_file) as f:
+                pdl_module_str = f.read()
+                parser = Parser(ctx, pdl_module_str)
+                pdl_module = parser.parse_module()
+        else:
+            pdl_module = payload_module
+        stream = StringIO()
+        rewrite_patterns = [
+            cast(RewritePattern, PDLRewritePattern(op, ctx, stream))
```

Comment:
is this cast necessary?

Reply:
Yeah because ```list``` is invariant so ```list[PDLRewritePattern]``` cannot be seen as ```list[RewritePattern]``` (```pyright``` error)

Reply:
Ah, but if you put it inline, it will work, right?

Reply:
> Ah, but if you put it inline, it will work, right?

Right ^^ It is done !